src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs

```diff
@@ -0,0 +1,231 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.Extensions.Internal;
+
+namespace System.Net
+{
+    [EventSource(Name = "System.Net.NameResolution")]
+    internal sealed class NameResolutionTelemetry : EventSource
+    {
+        public static readonly NameResolutionTelemetry Log = new NameResolutionTelemetry();
+
+        public static new bool IsEnabled => Log.IsEnabled();
+
+        private const int ResolutionStartingEventId = 1;
+        private const int ResolutionSuccessEventId = 2;
+        private const int ResolutionFailureEventId = 3;
+
+        private PollingCounter? _lookupsRequestedCounter;
+        private EventCounter? _lookupsDuration;
+
+        private long _lookupsRequested;
+
+        protected override void OnEventCommand(EventCommandEventArgs command)
+        {
+            if (command.Command == EventCommand.Enable)
+            {
+                // The cumulative number of name resolution requests started since events were enabled
+                _lookupsRequestedCounter ??= new PollingCounter("dns-lookups-requested", this, () => Interlocked.Read(ref _lookupsRequested))
+                {
+                    DisplayName = "DNS Lookups Requested"
+                };
+
+                _lookupsDuration ??= new EventCounter("dns-lookups-duration", this)
+                {
+                    DisplayName = "Average DNS Lookup Duration",
+                    DisplayUnits = "ms"
+                };
+            }
+        }
+
+
+        private const int MaxIPFormattedLength = 128;
+
+        // Methods below assume that ResolutionSuccess and ResolutionFailure have the same signature
+
+        [Event(ResolutionStartingEventId, Level = EventLevel.Informational)]
+        private void ResolutionStarting(string hostNameOrAddress) => WriteEvent(ResolutionStartingEventId, hostNameOrAddress);
+
+        [Event(ResolutionSuccessEventId, Level = EventLevel.Informational)]
+        private void ResolutionSuccess(string hostNameOrAddress, double duration) => WriteEvent(ResolutionSuccessEventId, hostNameOrAddress, duration);
+
+        [Event(ResolutionFailureEventId, Level = EventLevel.Informational)]
+        private void ResolutionFailure(string hostNameOrAddress, double duration) => WriteEvent(ResolutionFailureEventId, hostNameOrAddress, duration);
+
+
+        [NonEvent]
+        public ValueStopwatch BeforeResolution(string hostNameOrAddress)
+        {
+            Debug.Assert(hostNameOrAddress != null);
+
+            if (IsEnabled())
+            {
+                Interlocked.Increment(ref _lookupsRequested);
+
+                if (IsEnabled(EventLevel.Informational, EventKeywords.None))
+                {
+                    ResolutionStarting(hostNameOrAddress);
+                    return ValueStopwatch.StartNew();
```

Comment:
I'm skeptical that the stopwatch should be conditional on being at the informational level. The usual pattern elsewhere is that counters are tracked as long as the EventSource is enabled. I'd anticipate that if dotnet-counters tool monitored this code it report the value of duration counter is always zero because the EventSource would not have Informational level enabled.

Reply:
Good catch, this is my mistake after adding the `IsEnabled(EventLevel.Informational)`, the stopwatch should be returned as long as EventSource is enabled.